### PR TITLE
Fix YouTube Livestreams

### PR DESCRIPTION
**Pull request by FireController1847**
YouTube Livestreams currently start very early in the livestream, due to the way that YouTube's m3u8 works. This will cause it to read the stream from the ~4h mark (if the stream is archived). 

Fixes https://github.com/fent/node-ytdl-core/issues/157
Fixes https://github.com/fent/node-ytdl-core/issues/263

### CODE REVIEW

**Comment by FireController1847** [comment]
This is my current prototype of a working solution. Please edit it and make it match whatever you'd like.

Example usage:
```diff
# In line 68 on fent/node-ytdl-core, make these changes.
*  if (format.live) {
*    var req = m3u8stream(url, {
*      chunkReadahead: +info.live_chunk_readahead,
*      requestOptions: options.requestOptions,
+      ytdlOptions: {isYtdlCore: true}
*   });
*   req.on('error', stream.emit.bind(stream, 'error'));
*   stream.destroy = req.end.bind(req);

File: lib/m3u8-parser.js [hunk 1]
```diff
@@ -22,6 +27,26 @@ var m3u8parser = module.exports = function() {
...
+    var line = self.ytdlSavedLines[self.ytdlSavedLines.length-1];
+    if (!line) {
+      self.emit('ytdl-error', new Error('Line is null/undefined.\n' + time));
+      return;
+    }
+    parseLine(line);
+    var timeout = setTimeout(function() {
+      self.ytdlSavedLines = [];
+      return self.emit('ytdl-end');
+      clearTimeout(timeout); // Just in case timeout is still set, we want to make sure it's destroyed.
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: I do realize this line is now redundant, I added it when there wasn't a return for the self.emit. I was having issues in this area, though, with it requesting 4-5 times a second. I don't know if this still happens, but I think it is happening very slowly. The longer you listen, the more you hear stuttering. Try adding a console.log here to see how many times it calls. It should only all every 5 seconds (or however long <time> is).

TL;DR: I believe I ended up creating a memory leak right here.

File: lib/index.js [hunk 2]
```diff
@@ -56,12 +57,16 @@ module.exports = function(playlistURL, options) {
   }
 
   var tid;
+  var rpts;
```
Comment (FireController1847) [inline lib/index.js]: rpts means "Repeat Times". It's the amount of times refreshPlaylist() has been called when ytdl-error runs. If it loops more than 5 times from ytdl-errors in a row, then it's probably crashing and time to stop.

File: lib/index.js [hunk 3]
```diff
@@ -56,12 +57,16 @@ module.exports = function(playlistURL, options) {
...
+  var ytft = true;
```
Comment (FireController1847) [inline lib/index.js]: ytft means "YouTube First Time". This is required, as we have to let the m3u8 know if it's the first time. More information on this will be in my comment on the actual part that handles time.

File: lib/index.js [hunk 4]
```diff
@@ -74,6 +79,21 @@ module.exports = function(playlistURL, options) {
       totalItems++;
       requestQueue.push(item, onQueuedEnd);
     });
+    parser.on('ytdl-end', function() {
```
Comment (FireController1847) [inline lib/index.js]: We do not want to call the normal end, as that sets a 10 minute timeout, when we need to call every 5 seconds. So, this custom function will do just that. It deletes the parser just in case that causes a memory leak (because we create a new one every loop and don't "return"), and then calls refreshPlaylist again. It finally sets ytft to false and rpts to 0 (see above).

File: lib/index.js [hunk 5]
```diff
@@ -74,6 +79,21 @@ module.exports = function(playlistURL, options) {
...
+      fetchingPlaylist = false;
+      delete parser;
+      refreshPlaylist();
+      rpts = 0;
+      ytft = false;
+    });
+    parser.on('ytdl-error', function() {
```
Comment (FireController1847) [inline lib/index.js]: When the parser has an internal error when parsing time or something like that, this will be called. If rpts (see above) is more than or equal to 5, then error out and stop with 'Too Many Requests' signaling an internal error. Optionally, we might want to add a second argument called "err" as when this event is called it does come with reasoning.

File: lib/m3u8-parser.js [hunk 6]
```diff
@@ -8,9 +8,14 @@ const util     = require('util');
  * @extends WritableStream
  * @constructor
  */
-var m3u8parser = module.exports = function() {
+var m3u8parser = module.exports = function(ytdlOptions) {
   var lastLine = '';
   var self = this;
+  self.ytdlSavedLines = [];
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: We must save the lines from the data bus, as we need them ALL if we're ever going to add backwards-time frames as an option.

File: lib/m3u8-parser.js [hunk 7]
```diff
@@ -8,9 +8,14 @@ const util     = require('util');
...
+  if (ytdlOptions) {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Sets the YTDL options coming from the initial creation. We use these to determine if we need ytdl or not, and if it's the first time or not (see below).

File: lib/m3u8-parser.js [hunk 8]
```diff
@@ -22,6 +27,26 @@ var m3u8parser = module.exports = function() {
       self.emit('item', line.trim());
     }
   }
+  function ytdlParse() {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: A custom function to parse on exit. We could optionally have this literally just be in the event, but this function seemed like the better option. Descriptions of each item in the function below.

File: lib/m3u8-parser.js [hunk 9]
```diff
@@ -22,6 +27,26 @@ var m3u8parser = module.exports = function() {
...
+    var time = self.ytdlSavedLines[self.ytdlSavedLines.length - 2];
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: We have to get the time from the line, to determine how long the m3u8 packet is. Is it 5 seconds? 4 seconds? This will automatically get it for us. For every m3u8 item that comes through, right above it in the array is how long that item is, determining how long we should time out before asking for more information.

File: lib/m3u8-parser.js [hunk 10]
```diff
@@ -22,6 +27,26 @@ var m3u8parser = module.exports = function() {
...
+    if (!time) {
+      self.emit('ytdl-error', new Error('Time is null/undefined.\n' + time));
+      return;
+    }
+    time = parseInt(time.replace(/#EXTINF:/g, "").replace(/,/g, ""));
+    if (self.isFirstTime) time = 1;
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: When initially starting this loop, it will wait 5 seconds before actually outputting a stream, so we check if it's the first time so it can begin streaming immediately instead of waiting for the second packet. This will wait 1 second, to make sure the packet is parsed correctly and to give us time to live.

File: lib/m3u8-parser.js [hunk 11]
```diff
@@ -22,6 +27,26 @@ var m3u8parser = module.exports = function() {
...
+    var line = self.ytdlSavedLines[self.ytdlSavedLines.length-1];
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Get's the last line in the list. This is the most important part. After every packet that is sent in from the data, the last one in the array is the "live" value. The top one in the array (ignoring all the tags that are at the top) are actually the archive (hence why it started at -4hours when streaming originally). If we want to have future options to start at a certain time, this would be the place to code it.

File: lib/m3u8-parser.js [hunk 12]
```diff
@@ -22,6 +27,26 @@ var m3u8parser = module.exports = function() {
...
+    if (!line) {
+      self.emit('ytdl-error', new Error('Line is null/undefined.\n' + time));
+      return;
+    }
+    parseLine(line);
+    var timeout = setTimeout(function() {
+      self.ytdlSavedLines = [];
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Reset the array that way when the next packet comes in we're not filling the array with the same useless junk as before. Keeps the array updated and makes sure that no old information is kept.

File: lib/m3u8-parser.js [hunk 13]
```diff
@@ -30,17 +55,35 @@ var m3u8parser = module.exports = function() {
   this._write = function(chunk, encoding, callback) {
     var lines = chunk.toString('utf8').split('\n');
     if (lastLine) { lines[0] = lastLine + lines[0]; }
-    lines.forEach(function(line, i) {
-      if (i < lines.length - 1) {
-        parseLine(line);
-      } else {
-        // Save the last line in case it has been broken up.
-        lastLine = line;
+    if (this.isYtdlCore) {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Add a check for if you want the old functionality or the YouTube functionality. Might want to rename the variable, as I'm sure this isn't _just_ used in YtdlCore.

File: lib/m3u8-parser.js [hunk 14]
```diff
@@ -30,17 +55,35 @@ var m3u8parser = module.exports = function() {
...
+      for (var i = 0; i < lines.length; i++) {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Synchronously loop through all of the lines instead of Asynchronously, that way we are sure that all of the packets got into the array before the 'end' event is called from the ReadableStream.

File: lib/m3u8-parser.js [hunk 15]
```diff
@@ -30,17 +55,35 @@ var m3u8parser = module.exports = function() {
...
+        var line = lines[i];
+        if (i < lines.length - 1) {
+          self.ytdlSavedLines.push(line);
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Push every single packet into the array individually, instead of pushing sections of packets into the array.

File: lib/m3u8-parser.js [hunk 16]
```diff
@@ -30,17 +55,35 @@ var m3u8parser = module.exports = function() {
...
+        } else {
+          // Save the last line in case it has been broken up.
+          lastLine = line;
+        }
       }
-    });
+    } else {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Keep the old functionality just in case somebody wants to use this for something other than Ytdl-Core. m3u8 is very complicated.

File: lib/m3u8-parser.js [hunk 17]
```diff
@@ -30,17 +55,35 @@ var m3u8parser = module.exports = function() {
...
+        } else {
+          // Save the last line in case it has been broken up.
+          lastLine = line;
+        }
+      });
+    }
     callback();
   };
 
+  this.on('end', function() {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: When the end event is called, due to the fact we synchronously pushed the packets into the array we can be certain that every packet is now in the array, and we can now parse it manually using the function (of course, checking if isYtdlCore is true first).

File: lib/m3u8-parser.js (hunk 11 above)
Comment (FireController1847) [reply lib/m3u8-parser.js]: I'd like to just state that right now, I'm realizing we could optionally add 5 second time frames that we can go back in time in. The lower this number, the more 5 seconds we go back in time. So instead of if -1 for the last item (live), we could do -24 for 2 minutes back in the video, and so forth. The limit would have to be -14400, though, as that's four hours back in time (and cannot be guaranteed).

**Comment by FireController1847** [comment]
I'm pretty sure https://github.com/fent/node-m3u8stream/pull/7/files#diff-0dd6548c03db28e92002c355f52d343bL50 is causing the memory leak. I'm going to wait to continue on this project until you've commented.

**Comment by FireController1847** [comment]
To be honest, this was mostly just to demonstrate what we had to do to get it to actually work, so you could understand what I was talking about (because through text none of that would have made sense lol). I'm not exactly sure how you want to do this specifically.

**Comment by fent** [comment]
Yeah, I think I understand now. Thanks for taking the time to put this together and explain.

**Comment by FireController1847** [comment]
@fent Yeah no problem! This was definitely a learning process for me haha! :D If you need me to explain anything else about it just ask ;P

**Comment by FireController1847** [comment]
_waits patiently to see if he forgets about this_

**Comment by fent** [comment]
I haven't forgotten :)

**Comment by FireController1847** [comment]
So I've decided to take this on again, but this time in a production environment. I have a question, though. I'm not sure what to classify the option as because it's really indescribable. Possibly, "youtubeFormatting" or something similar? I'm just going to use "youtubeFormatting" for my current tests until I get a response :)

**Comment by FireController1847** [comment]
Okay, so I've made some changes and I've come to this. This is actually working very well as of right now, but there is one issue where it will just cut off (stop working) over a few seconds. It's working extremely well, though, and is much better than my previous commit.

To enable this mode, all you add are two options (using YTDL as the example):
```diff
    let req = m3u8stream(url, {
      chunkReadahead: +info.live_chunk_readahead,
      requestOptions: options.requestOptions,
+     refreshInterval: 4000,
+     youtubeFormatting: true
    });
```

File: lib/index.js [hunk 18]
```diff
@@ -61,7 +61,7 @@ module.exports = (playlistURL, options) => {
     fetchingPlaylist = true;
     var req = miniget(playlistURL, requestOptions);
     req.on('error', onError);
-    var parser = req.pipe(new m3u8());
+    var parser = req.pipe(new m3u8(options));
```
Comment (FireController1847) [inline lib/index.js]: Pass the options through, as they're needed for youtubeFormatting detection

File: lib/m3u8-parser.js [hunk 19]
```diff
@@ -10,8 +10,11 @@ const Writable = require('stream').Writable;
  * @constructor
  */
 module.exports = class m3u8parser extends Writable {
-  constructor() {
+  constructor(options) {
     super({ decodeStrings: false });
+    this.options = options;
+    this._firstLines = [];
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: We must save _firstLines so we output the initial tags.

File: lib/m3u8-parser.js [hunk 20]
```diff
@@ -10,8 +10,11 @@ const Writable = require('stream').Writable;
...
+    this.first = true;
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: This is required so we know which lines are _firstLines

File: lib/m3u8-parser.js [hunk 21]
```diff
@@ -34,14 +37,33 @@ module.exports = class m3u8parser extends Writable {
   _write(chunk, encoding, callback) {
     var lines = chunk.toString('utf8').split('\n');
     if (this._lastLine) { lines[0] = this._lastLine + lines[0]; }
-    lines.forEach((line, i) => {
-      if (i < lines.length - 1) {
-        this._parseLine(line);
-      } else {
-        // Save the last line in case it has been broken up.
-        this._lastLine = line;
+    if (!this.options.youtubeFormatting) {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: If we do not want to use youtubeFormatting, use the old way

File: lib/m3u8-parser.js [hunk 22]
```diff
@@ -34,14 +37,33 @@ module.exports = class m3u8parser extends Writable {
...
+      lines.forEach((line, i) => {
+        if (i < lines.length - 1) {
+          this._parseLine(line);
+        } else {
+          // Save the last line in case it has been broken up.
+          this._lastLine = line;
+        }
+      });
+    } else {
+      this._lastLine = lines[lines.length - 2];
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Set the last line to before the blank output [' '], which is two back. This is the last audio output.

File: lib/m3u8-parser.js [hunk 23]
```diff
@@ -34,14 +37,33 @@ module.exports = class m3u8parser extends Writable {
...
+    if (this.first) {
+      this._firstLines = lines;
+      this.first = false;
+    }
+    callback();
+  }
+
+  _final(callback) {
+    if (this.options.youtubeFormatting) {
+      const firstIndex = this._firstLines.findIndex(a => a.includes('#EXTINF'));
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Find the first audio mark in _firstLines

File: lib/m3u8-parser.js [hunk 24]
```diff
@@ -34,14 +37,33 @@ module.exports = class m3u8parser extends Writable {
...
+      for (let i = 0; i < firstIndex.length - 1; i++) {
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Loop through everything before the first audio mark so we get all of those juicy details about the stream

File: lib/m3u8-parser.js [hunk 25]
```diff
@@ -34,14 +37,33 @@ module.exports = class m3u8parser extends Writable {
...
+        this._parseLine(this._firstLines[i]);
       }
-    });
+      this._parseLine(this._lastLine);
```
Comment (FireController1847) [inline lib/m3u8-parser.js]: Parse the final audio line

**Comment by FireController1847** [comment]
@fent Any particular status on this new commit?

**Comment by fent** [comment]
Hi, sorry, I haven't had the time to look at this in detail.

**Comment by FireController1847** [comment]
Ah, that's okay, I fully understand.

**Comment by FireController1847** [comment]
Um?

**Comment by fent** [comment]
Sorry, was changing some things around. But don't worry, I still have this PR/feature in mind

**Comment by FireController1847** [comment]
Alright, cool. Just let me know if you need anything more explained!

**Comment by FireController1847** [comment]
@fent Sorry I'm pestering you about this, but it's been ~4 months since I made this PR with no significant changes. I'd love to be able to help out, but it's not really clear what is needed to do so (considering this was closed). This is a bug that many Discord.js users are hoping will get fixed, as it causes unstability in their programs. If there's anything you need me to do, please tell me, as I want to get this fixed as fast as possible.

**Comment by fent** [comment]
> it's been ~4 months since I made this PR with no significant changes

Sorry for the big delay. It's cool to feel the need to alert me. I'll try to prioritize this when I next dive into my open source stuff.

**Comment by FireController1847** [comment]
Alright cool. Thanks 👍

**Comment by FireController1847** [comment]
@fent This is your 3-monthly reminder to not forget about this PR! :P

**Comment by FireController1847** [comment]
@fent This is your do-it-when-a-reasonable-amount-of-time-has-passed reminder to not forget about this PR

**Comment by fent** [comment]
thesadfrog

**Comment by Gymnophoria** [comment]
@fent  ily  
_but_ it's almost the one year anniversary of this issue (56 more days 'till then lol) and i thought id give the reminder this time ['cause ive had this one card in my trello](https://trello.com/c/8hBIgjPL/95-a-command-for-play-stream) waiting around for a long time lmao

pls ;-;

we should throw a party at the anniversary 🤔

**Comment by FireController1847** [comment]
Holy fek, it is almost a year old.

**Comment by fent** [comment]
Sorry for the big delay. One of my goals as a dev is to not neglect any of  my active repos. Which I've failed to do here :(

I'll try to look at this at the beginning of next week.

**Comment by FireController1847** [comment]
Awesome, can't wait. Glad to hear you haven't _entirely_ forgotten about it :)

**Comment by FireController1847** [comment]
Woohoo! Glad we finally got it done @fent. Good on ya' 🎉
